abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+- `Done` means that data chunk should be forwarded upstream and that this
+  extension shouldn't be called again for the current stream.
+- `Pause` means that further processing should be paused until
+  `proxy_resume_stream` is called.
+- `WaitForMoreData` means that further processing should be paused and that host
+  should call this callback again when it has more data.
+- `WaitForEndOrFull` means that further processing should be paused and that
+  host should call this callback again when downstream connection is closed or
+  when its internal buffers are full.
+- `Close` means that this connection should be closed immediately.
+
+#### `proxy_on_downstream_close`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (proxy_close_source_type_t) close_source`
+* returns:
+  - none
+
+Called when the downstream connection (`stream_id`) is closed.
+
+#### `proxy_on_upstream_data`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) data_size`
+  - `i32 (bool) end_of_stream`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called for each data chunk received from upstream (`stream_id`).
+
+Data (of size `data_size`) can be retrieved using `proxy_get_buffer`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that data should be forwarded downstream.
+- `EndStream` means that data should be forwarded downstream and that this
+  connection should be closed afterwards.
+- `Done` means that data should be forwarded downstream and that this extension
+  shouldn't be called again for the current stream.
+- `Pause` means that further processing should be paused until
+  `proxy_resume_stream` is called.
+- `WaitForMoreData` means that further processing should be paused and that host
+  should call this callback again when it has more data.
+- `WaitForEndOrFull` means that further processing should be paused and that
+  host should call this callback again when when upstream connection is closed
+  or when its internal buffers are full.
+- `Close` means that this connection should be closed immediately.
+
+#### `proxy_on_upstream_close`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (proxy_close_source_type_t) close_source`
+* returns:
+  - none
+
+Called when the upstream connection (`stream_id`) is closed.
+
+### Functions exposed by the host environment
+
+#### `proxy_resume_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused downstream or upstream (`stream_type`) connection
+in current context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream wasn't paused.
+
+#### `proxy_close_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Close downstream or upstream (`stream_type`) connections in current context.
```

Comment:
see comments above about Envoy close types

Reply:
And if we have this why do we need the return codes for end stream, etc.? Can we just have a stop iteration return code?